src/symbol.rs

```diff
@@ -0,0 +1,345 @@
+        let mut syms = HashMap::new();
+        for lurk_sym in LURK_SYMBOLS {
+            syms.insert(Self::lurk_sym(&format!("{}", lurk_sym)), *lurk_sym);
+        }
+        syms
+    }
+
+    pub fn is_whitespace(c: char) -> bool {
+        LURK_WHITESPACE.iter().any(|x| *x == c)
+    }
+
+    pub fn escape_symbol_element(xs: &str) -> String {
+        let mut res = String::new();
+        for x in xs.chars() {
+            if ESCAPE_CHARS.chars().any(|c| c == x) {
+                res.push_str(&format!("\\{}", x));
+            } else if Self::is_whitespace(x) {
+                res.push_str(&format!("{}", x.escape_unicode()));
+            } else {
+                res.push(x)
+            }
+        }
+        res
+    }
+
+    pub fn print_root(&self) -> String {
+        format!("#{}", self.marker())
+    }
+
+    pub fn print_escape(&self) -> String {
+        if self.is_root() {
+            return self.print_root();
+        }
+        let mut res = String::new();
+        let xs = self.path();
+        if Self::sym_needs_marker(self) {
+            res.push(self.marker())
+        }
+        res.push_str(&Self::escape_symbol_element(&xs[0]));
+        for x in xs[1..].iter() {
+            res.push(SYM_SEPARATOR);
+            res.push_str(&Self::escape_symbol_element(x));
+        }
+        if xs[xs.len() - 1].is_empty() {
+            res.push(SYM_SEPARATOR);
+        }
+        res
+    }
+
+    pub fn sym_needs_marker(&self) -> bool {
+        let xs = self.path();
+        if self.is_root()
+            || self.is_key()
+            || xs[0].is_empty()
+            || xs[0] == "#"
+            || xs[0] == "~"
+            || xs[0] == self.print_root()
+        {
+            return true;
+        };
+        let c = xs[0].chars().next().unwrap();
+        "~#1234567890.:'[](){},\"\\".chars().any(|x| x == c)
+            || char::is_whitespace(c)
+            || char::is_control(c)
+    }
+}
+
+impl fmt::Display for Symbol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Symbol::Sym(_) if self.is_root() => {
+                write!(f, "~()")
+            }
+            Symbol::Key(_) if self.is_root() => {
+                write!(f, "~:()")
+            }
+            _ => write!(f, "{}", self.print_escape()),
+        }
+    }
+}
+
+impl From<&str> for Symbol {
+    fn from(s: &str) -> Symbol {
+        let sym_path = &s[1..s.len()].split(SYM_SEPARATOR).collect::<Vec<&str>>();
+        if s.starts_with(SYM_MARKER) {
+            Self::sym(sym_path.clone())
+        } else if s.starts_with(KEYWORD_MARKER) {
+            Self::key(sym_path.clone())
+        } else {
+            panic!("Invalid starter char for Symbol")
+        }
+    }
+}
+
+impl From<String> for Symbol {
+    fn from(s: String) -> Symbol {
+        (&s as &str).into()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Copy)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Arbitrary))]
+pub enum LurkSym {
```

Comment:
This probably should not survive once we have proper package handling. Something like it *might* be useful for use in the implementation, but it shouldn't be needed for most purposes. The only reason we should be special-casing Lurk symbols in the end is to the extent it's a useful optimization (for example, for circuit constants, etc.)

Reply:
acknowledged & agreed